packages/flutter_tools/lib/src/commands/attach.dart

```diff
@@ -114,14 +125,17 @@ class AttachCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    final String ipv4Loopback = InternetAddress.loopbackIPv4.address;
+    final String ipv6Loopback = InternetAddress.loopbackIPv6.address;
+
     Cache.releaseLockEarly();
 
     await _validateArguments();
 
     writePidFile(argResults['pid-file']);
 
     final Device device = await findTargetDevice();
-    final int devicePort = debugPort;
+    final int devicePort = debugPort ?? await MDnsObservatoryPortDiscovery().queryForPort(applicationId: appId);
```

Comment:
Check if the device is an iOSDevice (or emulator?) first

Reply:
I considered that but I like the idea of this "just working" once we wire up the needed bits in Android.

Most people should be passing debugPort anyway today.

Reply:
Fuchsia uses a different discovery mechanism, so no port is usually specified.

Reply:
I made this opt in for iOS devices and simulators only.  If/when we land support for Android or other platforms we can update.